test/test_sparse.py

```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
+                # Check sparse softmax definition
+
+                # check Python sparse softmax
+                y = sparse_softmax(x, dim)
+                r1 = softmax_to_dense(x, dim)
+                r2 = y.to_dense()
+                self.assertEqual(r1, r2)
+
+                # check C++ sparse softmax
+                y1 = torch.sparse.softmax(x, dim)
+                self.assertEqual(y, y1)
+
+                # check C++ sparse log_softmax
+                ly1 = torch.sparse.log_softmax(x, dim)
+                self.assertEqual(ly1, sparse_log(y1))
+
+                # Check autograd support on sparse softmax
+
+                # check softmax Jacobian definition for dense input
+                x1 = to_dense(x, fill_value='-inf')
+                J = softmax_jacobian_analytic(x1, dim)
+                assert J.shape[0] == x.shape[dim]
+                assert J.shape[dim + 1] == x.shape[dim]
+
+                # check softmax Jacobian from autograd, dense input
+                J2 = softmax_jacobian_autograd(x1, dim)
+                self.assertEqual(J, J2)
+
+                # check softmax Jacobian from autograd, sparse input
+                J3 = softmax_jacobian_autograd(x, dim)
+                self.assertEqual(J, J3)
+
+                '''
+                y = softmax(x, dim)
+                z = log(y) = log_softmax(x, dim)
+                Dy/Dx = J
+                Dz/Dx = Dz/Dy Dy/Dx = 1/y * J
+                => J = J_log * y
+                '''
+                # log_softmax Jacobian from autograd, dense input
+                J2_log = softmax_jacobian_autograd(x1, dim, log=True)
+
+                # log_softmax Jacobian from autograd, sparse input
+                J3_log = softmax_jacobian_autograd(x, dim, log=True)
+
+                J = J.transpose(0, dim + 1)
+                J2_log = J2_log.transpose(0, dim + 1)
+                J3_log = J3_log.transpose(0, dim + 1)
+                self.assertEqual(J, J2_log * r1)
+                self.assertEqual(J, J3_log * r1)
+
+                if dim == 0:
+                    # check dtype argument
+                    other_dtype = torch.float32
+                    y2 = torch.sparse.softmax(x, dim, dtype=other_dtype)
+                    self.assertEqual(y2.dtype, other_dtype)
+                    self.assertEqual(y2, y1.type(other_dtype))
+
+                    ly2 = torch.sparse.log_softmax(x, dim, dtype=other_dtype)
+                    self.assertEqual(ly2.dtype, other_dtype)
+                    self.assertEqual(ly2, ly1.type(other_dtype))
+
+        test_op(1, 10, [3])
+        test_op(1, 10, [2, 3])
+        test_op(1, 10, [3, 2])
+        test_op(2, 10, [2, 3, 4])
+        test_op(2, 10, [3, 4])
+        test_op(2, 5, [5, 4])
+        test_op(2, 10, [3, 4, 2])
+        test_op(3, 10, [3, 4, 2])
+        test_op(3, 100, [3, 4, 2])
+        test_op(3, 100, [3, 4, 2, 3])
+        test_op(3, 100, [3, 4, 2, 3, 5, 2])
+        test_op(4, 100, [3, 4, 2, 3, 5, 2])
```

Comment:
How long do these tests take to run?

Reply:
A single run of `test_softmax` takes about 0.5 seconds.